overlord/snapstate/snapstate.go

```diff
@@ -2116,6 +2116,18 @@ func doUpdate(st *state.State, requested []string, updates []update, opts Option
 			continue
 		}
 
+		if up.SnapState.RefreshFailures != nil {
+			// Check if snap revision is known to fail and if the current refresh needs to be skipped.
+			if up.SnapState.RefreshFailures.ToRevision != up.Setup.Revision() {
+				// Snap has new target revision not known to fail, let's reset RefreshFailures
+				// and continue refresh normally.
+				up.SnapState.RefreshFailures = nil
+				Set(st, up.SnapState.InstanceName(), &up.SnapState)
+			} else if shouldSkipSnapRefresh(&up.SnapState, up.Setup.Revision(), opts) {
+				continue
+			}
+		}
+
```

Comment:
Could this be simplified to this? There is already a `nil` check in shouldSkipSnapRefresh, and the check for revision.

```suggestion
		if shouldSkipSnapRefresh(&up.SnapState, up.Setup.Revision(), opts) {
		       continue
		} else if (up.SnapState.RefreshFailures != nil)  {
			// Snap has new target revision not known to fail, let's reset RefreshFailures
			// and continue refresh normally.
			up.SnapState.RefreshFailures = nil
			Set(st, up.SnapState.InstanceName(), &up.SnapState)
		}
```

Reply:
This is covered by https://github.com/canonical/snapd/pull/14578/files#diff-db72bd4eacd5472981461fb5bf207faae189669cde682db440c929328a9d13abR10850-R10858